// @un360/core/src/api/chatbot.ts
import type { ChatConfig, ProductContext, SuggestedQuestion, WidgetConfig } from '../types';

export class ChatbotAPI {
  private apiKey: string;
  private baseUrl: string;
  private productContext: ProductContext;
  private sessionId: string;

  constructor(config: ChatConfig) {
    this.apiKey = config.apiKey;
    
    // ✅ Robust fallback - NO process.env
    if (config.apiBaseUrl && typeof config.apiBaseUrl === 'string' && config.apiBaseUrl.trim() !== '') {
      this.baseUrl = config.apiBaseUrl.trim();
    } else {
      this.baseUrl = 'https://sdk-backend-teaa.onrender.com';
    }
    
    this.productContext = config.productContext;
    this.sessionId = config.sessionId || this.generateSessionId();
    
    console.log('ChatbotAPI baseUrl:', this.baseUrl);
  }

  private generateSessionId(): string {
    return `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  private async request<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`;
    console.log('Request URL:', url);
    
    const response = await fetch(url, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.apiKey,
        ...options.headers,
      },
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: 'Unknown error' }));
      throw new Error(error.error || `HTTP ${response.status}`);
    }

    return response.json();
  }

  async sendMessage(message: string): Promise<string> {
    const data = await this.request<{ response: string; session_id: string }>(
      '/api/v1/chat',
      {
        method: 'POST',
        body: JSON.stringify({
          message,
          product_context: this.productContext,
          session_id: this.sessionId,
        }),
      }
    );

    return data.response;
  }

  async getSuggestedQuestions(): Promise<SuggestedQuestion[]> {
    return this.request<SuggestedQuestion[]>('/api/v1/questions', {
      method: 'POST',
      body: JSON.stringify({
        sku: this.productContext.sku,
      }),
    });
  }

  async getWidgetConfig(): Promise<WidgetConfig> {
    return this.request<WidgetConfig>('/api/v1/config');
  }

  updateProductContext(newContext: Partial<ProductContext>) {
    this.productContext = { ...this.productContext, ...newContext };
  }
}
