// @un360/ui/src/components/ChatInput.tsx
import React, { useState } from 'react';

interface ChatInputProps {
  onSend: (message: string) => void;
  placeholder: string;
  theme: any;
}

export const ChatInput: React.FC<ChatInputProps> = ({ onSend, placeholder, theme }) => {
  const [inputValue, setInputValue] = useState('');

  const handleSend = () => {
    if (inputValue.trim()) {
      onSend(inputValue);
      setInputValue('');
    }
  };

  return (
    <div
      style={{
        padding: '16px',
        borderTop: '1px solid #e0e0e0',
        backgroundColor: theme.secondaryColor,
        display: 'flex',
        gap: '8px',
      }}
    >
      <input
        type="text"
        value={inputValue}
        onChange={(e) => setInputValue(e.target.value)}
        onKeyPress={(e) => e.key === 'Enter' && handleSend()}
        placeholder={placeholder}
        style={{
          flex: 1,
          padding: '10px 14px',
          border: '1px solid #e0e0e0',
          borderRadius: '20px',
          fontSize: '14px',
          outline: 'none',
        }}
      />
      <button
        onClick={handleSend}
        style={{
          backgroundColor: theme.primaryColor,
          color: theme.secondaryColor,
          border: 'none',
          borderRadius: '50%',
          width: '40px',
          height: '40px',
          cursor: 'pointer',
          fontSize: '18px',
        }}
      >
        ↑
      </button>
    </div>
  );
};
