// @un360/ui/src/components/ChatWidget.tsx
import React, { useState, useRef } from 'react';
import { useChat } from '@un360/core';
import { ChatMessages } from './ChatMessages';
import { ChatInput } from './ChatInput';
import { FloatingButton } from './FloatingButton';
import { useClickOutside } from '../hooks/useClickOutside';
import { ChatHeader } from './ChatHeader';

export interface ThemeConfig {
  primaryColor?: string;
  secondaryColor?: string;
  backgroundColor?: string;
  fontFamily?: string;
}

export interface ChatWidgetProps {
  apiKey: string;
  productId: string;
  apiBaseUrl?: string;
  position?: 'bottom-right' | 'bottom-left' | 'top-right' | 'top-left';
  theme?: ThemeConfig;
  greetingMessage?: string;
  placeholder?: string;
}

export const ChatWidget: React.FC<ChatWidgetProps> = ({
  apiKey,
  productId,
  apiBaseUrl,
  position = 'bottom-right',
  theme = {},
  greetingMessage = 'Hello! Ask me about this product.',
  placeholder = 'Type your message...',
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const [isMinimized, setIsMinimized] = useState(false);
  const [isMaximized, setIsMaximized] = useState(false);
  
  const chatRef = useRef<HTMLDivElement>(null);
  
  const {
    messages,
    sendMessage,
    isBotTyping,
    suggestedQuestions,
  } = useChat({ apiKey, productId, apiBaseUrl });

  useClickOutside(chatRef, () => setIsOpen(false));

  const defaultTheme = {
    primaryColor: theme.primaryColor || '#1976d2',
    secondaryColor: theme.secondaryColor || '#fff',
    backgroundColor: theme.backgroundColor || '#f5f5f5',
    fontFamily: theme.fontFamily || 'Arial, sans-serif',
  };

  const handleMinimize = () => {
    setIsMinimized(true);
    setIsMaximized(false);
  };

  const handleMaximize = () => {
    setIsMaximized(true);
    setIsMinimized(false);
  };

  const handleClose = () => setIsOpen(false);
  const toggleChat = () => setIsOpen(!isOpen);

  if (!isOpen) {
    return <FloatingButton onClick={toggleChat} position={position} theme={defaultTheme} />;
  }

  return (
    <div
      ref={chatRef}
      style={{
        position: 'fixed',
        [position.includes('right') ? 'right' : 'left']: '20px',
        [position.includes('bottom') ? 'bottom' : 'top']: '20px',
        width: isMaximized ? '600px' : '350px',
        height: isMinimized ? '60px' : isMaximized ? '80vh' : '500px',
        backgroundColor: defaultTheme.secondaryColor,
        borderRadius: '12px',
        boxShadow: '0 8px 24px rgba(0,0,0,0.2)',
        display: 'flex',
        flexDirection: 'column',
        overflow: 'hidden',
        transition: 'all 0.3s ease',
        fontFamily: defaultTheme.fontFamily,
        zIndex: 10000,
      }}
    >
      <ChatHeader
        onMinimize={handleMinimize}
        onMaximize={handleMaximize}
        onClose={handleClose}
        theme={defaultTheme}
      />

      {!isMinimized && (
        <>
          <ChatMessages
            messages={messages}
            isBotTyping={isBotTyping}
            greetingMessage={greetingMessage}
            suggestedQuestions={suggestedQuestions}
            onQuestionClick={(question) => sendMessage(question)}
            theme={defaultTheme}
          />
          
          <ChatInput
            onSend={sendMessage}
            placeholder={placeholder}
            theme={defaultTheme}
          />
        </>
      )}
    </div>
  );
};
