// @un360/ui/src/components/ChatHeader.tsx
import React from 'react';

interface ChatHeaderProps {
  onMinimize: () => void;
  onMaximize: () => void;
  onClose: () => void;
  theme: any;
}

export const ChatHeader: React.FC<ChatHeaderProps> = ({
  onMinimize,
  onMaximize,
  onClose,
  theme,
}) => {
  return (
    <div
      style={{
        backgroundColor: theme.primaryColor,
        color: theme.secondaryColor,
        padding: '16px',
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
      }}
    >
      <span style={{ fontWeight: 'bold', fontSize: '16px' }}>Product Assistant</span>
      <div style={{ display: 'flex', gap: '8px' }}>
        <button onClick={onMinimize} style={buttonStyle}>−</button>
        <button onClick={onMaximize} style={buttonStyle}>□</button>
        <button onClick={onClose} style={buttonStyle}>×</button>
      </div>
    </div>
  );
};

const buttonStyle = {
  background: 'transparent',
  border: 'none',
  color: 'white',
  fontSize: '20px',
  cursor: 'pointer',
  padding: '0 6px',
};
