// @un360/core/src/types/index.ts
export interface ChatMessage {
  sender: 'user' | 'chatbot';
  text: string;
  timestamp?: number;
}
export interface ProductContext{
  sku:string
  name:string 
  description?:string
  price?:number
  category?:string
  brand?:string
  specifications?:Record<string,any>
  images?:string[]
  url?:string
  inStock?:boolean
  [key:string]:any

}
export interface ChatConfig {
  apiKey: string;
  productContext:ProductContext,
  apiBaseUrl?: string;
  sessionId?: string;
}

export interface SuggestedQuestion {
  id: string;
  question: string;
}

export interface WidgetConfig {
  theme: {
    primary_color: string;
    secondary_color: string;
    background_color: string;
    font_family: string;
  };
  position: string;
  greeting_message: string;
  placeholder: string;
}

export interface UseChatOptions {
  apiKey: string;
  productContext:ProductContext
  apiBaseUrl?: string;
  onError?: (error: Error) => void;
}

export interface UseChatReturn {
  messages: ChatMessage[];
  sendMessage: (message: string) => Promise<void>;
  isLoading: boolean;
  isBotTyping: boolean;
  suggestedQuestions: SuggestedQuestion[];
  error: Error | null;
  clearMessages: () => void;
}
